Dynamic modules - throw on star exports

guybedford opened the pull request.

guybedford committed:
Normative: GetExportedNames adjustments for late-defined export names

This provides two features needed for supporting late-defined export
names on custom module records. Firstly, a second argument to
GetExportedNames of the namespace module being constructed by this call,
allowing it to be tracked, and secondly permitting ResolveExport
and GetExportedNames operations to provide abrupt completions
corresponding to instantiation errors.

guybedford committed:
instead of GetExportNames allow HostDefined on namespace objects

guybedford committed:
null return from GetExportedNames for disabling star reexports

guybedford committed:
tweak list for wording

guybedford committed:
preventExtensions invariant handling

guybedford committed:
clarify assertion

guybedford committed:
ban star reexports from dynamic modules

guybedford commented:
(the last commit is the new change)

littledan commented on spec.html:
```diff
@@ -8444,7 +8444,7 @@ <h1>Module Namespace Exotic Objects</h1>
             </td>
             <td>
-              List of String
+              List of String | *null*
```
Doesn't the need for all this | null/extensibility stuff go away?

littledan replied:
Ah, never mind, no, this is how CJS modules are directly (it is just never that a ESM is like that), right?

guybedford replied:
Yes, the pending state still stands - this is for the case on the 7th slide of https://docs.google.com/presentation/d/1--k5tJmhXJeLlUOmBdHGK3my9VCbZVL0MF8OrTeqS0I/edit?usp=sharing.

littledan commented on spec.html:
```diff
@@ -22034,6 +22048,7 @@ <h1>GetExportedNames ( _exportStarSet_ ) Concrete Method</h1>
           </emu-alg>
           <emu-note>
             <p>GetExportedNames does not filter out or throw an exception for names that have ambiguous star export bindings.</p>
+            <p>A *null* return from GetExportedNames indicates a non-Source-Text Module Record does not support star exports. This throws a corresponding SyntaxError exception.</p>
```
Nit: This might be clearer as a separate note

guybedford replied:
Sure, thanks.

guybedford committed:
update notes

guybedford commented:
I've added another commit to try and clarify the note cases here even further. Would value feedback on the wording here before updating the ECMA262 PR.

guybedford committed:
possible addition

guybedford commented:
I've added an additional commit to this at https://github.com/guybedford/ecma262/pull/3/commits/1b409d3eb0248770efd4abe0f5de917efc30c190, which is basically just reducing a minor spec complexity.

To explain:

* If the module hasn't executed yet, GetModuleNamespace calls GetExportedNames which must return *null* to go into the pending state, which then passes handling of the namespace finalization over to the dynamic modules spec.
* If the module has already executed though, GetModuleNamespace should do the finalization rather than the dynamic modules spec, so GetExportedNames in this case only should actually return the valid export names.
* This means that the definition of a "deferred exports" module isn't just that it returns *null* from GetExportedNames anymore, but that it returns *null* in the right cases (during star reexport lookups or for the module's own namespace when the module is unexecuted and detecting this difference based on the *ns* argument we added to GetExportedNames).

One way to make this a bit simpler would just be to always have the dynamic modules spec create the namespaces, and then to always create namespaces for all modules upfront instead of lazily. This is just in spec text though, so implementations could still make namespaces lazily in their own work if they want to.

The spec text then simplifies as provided in https://github.com/guybedford/ecma262/pull/3/commits/1b409d3eb0248770efd4abe0f5de917efc30c190 (no new argument needed for GetExportedNames, and *null* return happens consistently)

The dynamic modules proposal adjustment for this is at https://github.com/nodejs/dynamic-modules/pull/11/files, where namespaces are always finalized at https://github.com/nodejs/dynamic-modules/pull/11/files#diff-3540caefa502006d8a33cb1385720803R176.

Sorry to drag this one out, have been on and off it over the weekend - would value feedback on the above before a final merge!

guybedford committed:
rework star exports exception

jdalton commented:
I'm not a fan of making `export * from "cjs"` throw.
Please see my comment https://github.com/nodejs/dynamic-modules/pull/11#issuecomment-446476928.

guybedford commented:
As discussed in the modules meeting, I'm going ahead with merging this into the TC39 PR for dynamic modules, while we can continue to discuss alternative approaches.
